pinot-common/src/main/java/org/apache/pinot/common/utils/RegexpPatternConverterUtils.java

```diff
@@ -64,24 +71,42 @@ public static String likeToRegexpLike(String likePattern) {
         break;
     }
 
-    String escaped = escapeMetaCharacters(likePattern.substring(start, end));
-    StringBuilder sb = new StringBuilder(escaped.length() + 2);
+    likePattern = likePattern.substring(start, end);
+    StringBuilder sb = new StringBuilder();
     sb.append(prefix);
-    sb.append(escaped);
-    sb.append(suffix);
 
+    // handling SQL wildcards by replacing them with corresponding regex equivalents
+    // we ignore them if the SQL wildcards are escaped
     int i = 0;
-    while (i < sb.length()) {
-      char c = sb.charAt(i);
+    boolean isPrevCharBackSlash = false;
+    while (i < likePattern.length()) {
+      char c = likePattern.charAt(i);
       if (c == '_') {
-        sb.replace(i, i + 1, ".");
+        sb.append(isPrevCharBackSlash ? c : ".");
       } else if (c == '%') {
-        sb.replace(i, i + 1, ".*");
-        i++;
+        sb.append(isPrevCharBackSlash ? c : ".*");
```

Comment:
Why was this logic moved out of escapeMetaCharacters?

Reply:
So, escapeMetaCharacters was escaping every instance of a meta character. Now "\" was also a meta character, but we shouldn't escape that if that was used to escape the following character - like: \_, \%. Basically there are two flavours of "\" - one as a meta character, one as an escape character for the following character. We should escape the first and not the latter in the output. Since we can't replace every occurrence of "\" with "\\" - so I incorporated that logic here where if it's not before any SQL wildcard, we end up escaping it.

Reply:
That should not impact doing the processing in a separate method. I would still advise moving to a separate method to ensure a higher degree of readability.